cEP-0028.md

```diff
@@ -0,0 +1,444 @@
+# Optimizing coala for language server
+
+| Metadata |                                       |
+| -------- | ------------------------------------- |
+| cEP      | 0028                                  |
+| Version  | 0.1                                   |
+| Title    | Optimizing coala for language server  |
+| Authors  | Kilari Teja <mailto:ksdme9@gmail.com> |
+| Status   | Proposed                              |
+| Type     | Process                               |
+
+## Abstract
+
+This cEP describes the optimizations that need to be carried out
+to improve the performance of the coala based language server.
+
+## Introduction
+
+Language servers are the tools used to serve code editors with
+functionalities such as linting, code fixing, analysis etc in a
+standalone and independent manner. coala as a code style checker,
+linting tool and code fixer could be a powerful tool to be
+integrated into the code editor.
+
+Code Editing is a continuous process and the editors are expected
+to provide feedback on the fly, coala on the other hand was primarily
+designed as an analysis tool to be used after an incremental update to
+the codebase. Thus, the current implementation of a coala language
+server is severely limited by coala's performance.
+
+The performance of the server can be improved by primarily optimizing
+the coala and language server interface and improving the bear
+configuration options to avoid long running bears.
+
+## Proposal
+
+The optimization process focuses to maintain minimal changes to coala
+core and rather optimize the coala and language server interface. The
+following are the proposed ideas.
+
+1. Section Tagging
+
+   Some sections are not often required to be executed on every cycle of
+   coala analysis especially during use cases such as this. The parameter
+   set can thus be extended to support a parameter named `tags`. Since
+   these parameters are not associated with any particular bear but rather
+   control the effect of various sections during executing coala, no support
+   extension for bears would be necessary. These parameters enable support
+   for hooking a section for a particular tag type. For example:
+
+   ```
+   [longrunning]
+   tags = save
+   enabled = True
+   bears = SomeLongRunningBear
+   ```
+
+   This kind of a configuration will enable this section during analysis
+   cycles of coala via a language server when a save request is raised
+   while not impacting any untagged regular invocations.
+
+   ```
+   [longrunning]
+   files = *.py
+   bears = SomeLongRunningBear
+
+   [longrunning.some]
+   tags += save, change
+   bears = SomeLongRunningBear
+   ```
```

Comment:
Your tags are a good solution to the problem to make coala running smoother, however, can we maybe consider to tag certain bears by default without manual configuration? E.g. global bears generally tend to be slow, so they could be by default tagged `save` only. Or we allow bear devs to specify such default tags. Users then need less to configure, and coala works better out of the box. Just an idea :+1:

Reply:
@jayvdb also suggested the same idea of automatic tagging, he threw around the idea of using a `bears.json`. My only concern here is this was intended to offer a section based configuration mechanism rather than a bear based one because of some obvious fine tuning capabilities it provides over bears.

Reply:
This could however be made possible using hacky means by using `bears.json` and filtering bears from tagged sections  based upon their support for the tag type. This cannot be a long term solution as the tags may change over time and prevent tagging from being a general enough solution.

Reply:
Not bears.json but coala bears json. `coala -B --json` iirc.
coala-ls runtime decision making based on bear metadata at init time, esp. global

I think there will not be time for that now. We can add that later, or automatically allocate tags inside coala. Separate enhancement. Tags will suffice for now.